[AutoDiff upstream] Add `@transpose` attribute.

saeta opened the pull request:
The `@transpose(of:)` attribute registers a function as a transpose of another
function. This patch adds the `@transpose(of:)` attribute definition, syntax,
parsing, and printing.

Resolves TF-827.

---

Examples:

```swift
extension FloatingPoint where Self: Differentiable {
    @transpose(of: +)
    static func transposeAdd(_ v: Self) -> (Self, Self) { (v, v) }

    @transpose(of: *, wrt: 0)
    @transpose(of: *, wrt: 1)
    static func transposeMultiply(lhs: Self, rhs: Self) -> Self { lhs * rhs }
}
```

The [differentiable programming manifesto](https://github.com/apple/swift/blob/master/docs/DifferentiableProgramming.md#make-a-function-differentiable-using-derivative-or-transpose) has more information and examples.

---

Note: this code is upstreamed from `tensorflow` branch.

Known todos and issues are referenced in comments:
- Upstream `@transpose` attribute type-checking (TF-830, TF-1060).
- Enable `@transpose` attribute serialization (TF-838).
- Use module-qualified names instead of custom qualified name syntax/parsing
  (TF-1066).

We would like to upstream code and continue development on `master`, if that's acceptable!

dan-zheng committed:
[AutoDiff upstream] Add `@transpose(of:)` attribute.

The `@transpose(of:)` attribute registers a function as a transpose of another
function. This patch adds the `@transpose(of:)` attribute definition, syntax,
parsing, and printing.

Resolves TF-827.

Todos:
- Type-checking (TF-830, TF-1060).
- Enable serialization (TF-838).
- Use module-qualified names instead of custom qualified name syntax/parsing
  (TF-1066).

dan-zheng commented:
I revived this PR to upstream the current state of `tensorflow` branch, it is now ready for review. cc @rxwei @lattner 

The PR description tracks todos and known issues.

lattner commented on include/swift/AST/Attr.h:
```diff
@@ -1858,6 +1858,89 @@ class DerivativeAttr final
   }
 };
 
+/// The `@transpose` attribute registers a function as a transpose of another
```
nit: `@transpose(of:)`

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27545/commits/b94c097de935fd9a93b358aaf2d17f52c5e0b748.

lattner commented on test/AutoDiff/Parse/transpose_attr_parse.swift:
```diff
@@ -0,0 +1,92 @@
+func transpose(v: Float) -> Float
+
+// Qualified operator.
+// TODO(TF-1065): Consider disallowing qualified operators.
```
FWIW, I think that your implemented behavior is correct and nothing needs to change here.  "TF-1065" is really about the existing swift compiler being buggy and not allowing qualified names for operators like ..< . I'd suggest splitting that off to a swift bug, and I'd recommend that operator having to be spelled as Swift.`..<`

dan-zheng replied:
Thanks for the thorough review! I started a [discussion](https://forums.swift.org/t/operator-member-syntax/31773) regarding operator member syntax with proposed backtick syntax (`` SIMD2<Float>.`.<` ``).

As stated in TF-1065, qualified operator syntax isn't strictly necessary for `@transpose` or `@derivative` attribute because type-checking for these attributes require the operator to be defined in the same type context, where lookup is not ambiguous.

Taking action on these parser changes seems low priority so I'll leave the code and tests as is for now.

lattner reviewed (approved):
Nice!

dan-zheng committed:
Clarify `@derivative(of:)` and `@transpose(of:)` attribute names.

dan-zheng committed:
Fix qualified name parsing syntax verification errors.

dan-zheng committed:
Merge branch 'master' of github.com:apple/swift into upstream-transpose-attr

dan-zheng commented:
https://github.com/apple/swift-syntax/pull/189
@swift-ci Please test

dan-zheng commented:
CI failed for `test/IDE/complete_decl_attribute.swift` but the test passes for me locally.
Perhaps a clean test is necessary, I'll try it now.

dan-zheng commented:
apple/swift-syntax#189
@swift-ci Please clean test

rintaro commented on lib/Parse/ParseType.cpp:
```diff
@@ -633,12 +633,42 @@ ParserStatus Parser::parseGenericArguments(SmallVectorImpl<TypeRepr *> &Args,
+///   'Foo.f' -> true
+///   'Foo.Bar.f' -> true
+///   'f' -> false, no base type
+bool Parser::canParseBaseTypeForQualifiedDeclName() {
```
It would be great if  this can share some code with `canParseTypeIdentifier()`.
Also please move this function below line 1374.

dan-zheng replied:
Thanks for the catch!

I moved `Parser::canParseBaseTypeForQualifiedDeclName` just below `Parser::canParseTypeIdentifier` to line 1562. I also removed the doc comment on the function definition to be consistent with the surrounding code.

dan-zheng replied:
Code sharing done in https://github.com/apple/swift/pull/27545/commits/fa31c75aac4ac56c3b15e59b42bdc0c941cb2498.

I factored code into a new utility `Parser::canParseSimpleTypeIdentifier`, used by both `Parser::canParseTypeIdentifier` and `Parser::canParseBaseTypeForQualifiedDeclName`.

rintaro commented on test/AutoDiff/Parse/transpose_attr_parse.swift:
```diff
@@ -0,0 +1,92 @@
+func transpose(v: Float) -> Float
+
+// `subscript` keyword edge case.
+@transpose(of: Swift.Array.subscript(_:))
```
Could you add test cases for specialized generic types? Like: `@transpose(of: Foo<Int>.Bar.funcname)`

dan-zheng replied:
Specialized generic type test added in https://github.com/apple/swift/pull/27545/commits/c736938911f0d2c4b5fd08968c365bc58e1c0539. It passes, as expected.

rintaro commented on utils/gyb_syntax_support/AttributeNodes.py:
```diff
@@ -311,13 +314,56 @@
+                   ''',
+                   node_choices=[
+                       Child('MemberType', kind='MemberTypeIdentifier'),
+                       Child('SimpleType', kind='SimpleTypeIdentifier'),
```
I know the parser doesn't allow `@traspose(of: [Int].somehting)` But as a syntax declaration, I don't think need to limit this for type identifiers (i.e. we don't need `node_choices` at all).

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27545/commits/c736938911f0d2c4b5fd08968c365bc58e1c0539.

rintaro commented on utils/gyb_syntax_support/AttributeNodes.py:
```diff
@@ -311,13 +314,56 @@
+                       Child('PostfixOperator', kind='PostfixOperatorToken'),
+                       Child('SpacedBinaryOperator',
+                             kind='SpacedBinaryOperatorToken'),
+                   ]),
```
```python
    kind='Token',
    token_choices=[
        'IdentifierToken',
        'UnspacedBinaryOperatorToken',
        'SpacedBinaryOperatorToken',
        'PrefixOperatorToken',
        'PostfixOperatorToken',
    ]),
```
This should work. e.g. `FunctionDeclSyntax.Identifier`

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27545/commits/c736938911f0d2c4b5fd08968c365bc58e1c0539, thank you!

dan-zheng replied:
Would you like to be added as a reviewer on future parsing/syntax PRs? I'm happy to do so!

rintaro replied:
> Would you like to be added as a reviewer on future parsing/syntax PRs?

Yes. I appreciate it!

rintaro replied:
> Would you like to be added as a reviewer on future parsing/syntax PRs?

Yes. I appreciate it

dan-zheng replied:
Great, will do 🙂

dan-zheng committed:
Address parsing/syntax review feedback.

dan-zheng committed:
Create shared utility `Parser::canParseSimpleTypeIdentifier`.

Use it in both `Parser::canParseTypeIdentifier` and
`Parser::canParseBaseTypeForQualifiedDeclName`.

rintaro reviewed (approved):
Parser/Syntax changes look good!

dan-zheng commented:
https://github.com/apple/swift-syntax/pull/189
@swift-ci Please clean test
